### ISSUE TITLE
Creating a 'work' queue

### ISSUE THREAD

**Issue by davidchiles** [yapstudios/YapDatabase#235]
I wanted to use Yap to create a sort of persistent work queue.

The example I'm thinking of is a messaging app. Each message moves through multiple 'states' like being encrypted, sending to the server, and handling error cases.

So in this case the 'encryption listener' would be on the end of queue for messages needing encryption. Then once encrypted they would be stored back into yap and marked for needing sending. And then the 'sending listener' would pick up the message and send it over the wire. This helps us in the case where the device is offline so we don't loose any messages or are unsure of the messages state.

My question is what's the best yap tool for the job? It seems Views would be the closest even if they have a few other features that wouldn't be necessary. Or would it be worthwhile to create a new extension for this?

**Comment by robbiehanson** [comment]
Challenge accepted.

**Comment by robbiehanson** [comment]
Here you go:
https://github.com/yapstudios/YapDatabase/tree/experimental/action-manager/YapDatabase/Extensions/ActionManager

It's a new extension called YapDatabaseActionManager. Here's how it works:

Objects stored in the database can optionally implement the ```YapActionable``` protocol. This basically means implementing the ```- (NSArray<YapActionItem*> *)yapActionItems``` method, which allows the object to create and return zero or more associated YapActionItems.

The ```YapActionItem``` class encompasses all the information needed for the YapDatabaseActionManager to manage the action. Such as:
- **block** - action to be executed
- **date** - when the action should first be executed
- **requiresInternet** - whether to delay execution until Internet is available
- **retryTimeout** - whether to automatically retry, and how often to do so

The YDBActionManager handles all the associated logic of managing, executing & re-executing the actionItems. Which means you can simply specify a "block+date" of something you want done, and the system handles it for you.

Let's look at a quick example:

```objective-c
/**
 * This goes into one of your model objects.
**/
- (NSArray<YapActionItem*> *)yapActionItems
{
	YapActionItem *oneTimeAction = nil;
	YapActionItem *repeatingAction = nil;
	
	if (oneTimeActionNeeded)
	{
		YapActionItemBlock block = ^(NSString *collection, NSString *key, id object, id metadata){
			[MYAppDelegate doOneTimeAction:self.UUID];
		};
		
		oneTimeAction = [[YapActionItem alloc] initWithIdentifier:@"oneTimeAction"
		                                                     date:nil  // do it now
		                                             retryTimeout:60.0 // when to assume failure, and re-invoke block
		                                         requiresInternet:NO
		                                                    block:block];
	}
	if (nextRepeatingActionDate)
	{
		YapActionItemBlock block = ^(NSString *collection, NSString *key, id object, id metadata){
			[MYAppDelegate doRepeatingAction:self.UUID];
		};
		
		repeatingAction = [[YapActionItem alloc] initWithIdentifier:@"repeatingAction"
		                                                       date:nextRepeatingActionDate
		                                               retryTimeout:(60 * 5)
		                                           requiresInternet:YES // don't bother invoking block without Internet
		                                                      block:block];
	}
	
	if (oneTimeAction) {
		if (repeatingAction)
			return @[ oneTimeAction, repeatingAction ];
		else
			return @[ oneTimeAction ];
	}
	else {
		if (repeatingAction)
			return @[ repeatingAction ];
		else
			return nil;
	}
}
```

**oneTimeAction**:

This is an example of something that needs to be done once. For example, uploading something to a server. You can use appropriately named instance variables in your object to specify whether something needs to be done or not. (e.g. needsUploadAvatar)

The idea is, once the "oneTimeAction" has been completed, you do a readWriteTransaction, and unset the corresponding flag. (e.g. user.needsUploadAvatar = NO). YapDatabaseActionManager will see that you modified the object, and will re-query for the list of yapActionItems. And this time the "oneTimeAction" will be missing from the array. So the ActionManager will consider it completed. (And won't retry it anymore.)

**repeatingAction**:

This is an example of something you want to do every so often. For example, refreshing information from the server. Or deleting stuff that has expired.

The idea is, once the "repeatingAction" has been completed, you do a readWriteTransaction, and update the corresponding date. (e.g. user.nextRefreshDate = now + 7days) YapDatabaseActionManager will see that you modified the object, and will re-query for the list of yapActionItems. And this time the "repeatingAction" will be different. That is, it will have the same "identifier", but the date will be different. So the ActionManager will consider the previous action (same identifier, previous/old date) to be completed, and will schedule new new action (same identifier, new/future date).

 > It seems Views would be the closest even if they have a few other features that wouldn't be necessary.

YapDatabaseActionManager actually uses a View internally in order to sort all the objects in the database that have YapActionItems. This gives it the persistence required to scale to large database sizes.

Let me know if you have any questions.

PS - I've been using something similar to this in my own projects for some time. So it's somewhat tested. But I had to port it over to be an "official" YapDatabase utility. So let me know what you think.

**Comment by davidchiles** [comment]
Thanks! This looks really good. I'm away from my computer for a while. I'll take a closer look once I get back.

**Comment by robbiehanson** [comment]
Make sure you grab that fix ^

**Comment by davidchiles** [comment]
Having trouble building from the podspec because due to Reachability not included in podspec.

**Comment by davidchiles** [comment]
@robbiehanson Anything I can do to help get this merged in master?

**Comment by chrisballinger** [comment]
Hey @robbiehanson do you anticipate any significant changes to `YapDatabaseActionManager` before merging to master? We are planning to start writing some production code that uses this in the upcoming weeks.

Hope all is well!

**Comment by robbiehanson** [comment]
@davidchiles @chrisballinger

I'm already using it in production code.

> do you anticipate any significant changes

To be honest, I like the architecture, but I think it has one big thorn in it's side. I don't like these methods in the YapActionable protocol:
```objective-c
- (BOOL)hasYapActionItems;

- (NSDate *)earliestYapActionItemDate;
```

The only reason they exist is to speed up the groupingBlock & sortingBlock. But ultimately they make it really easy to shoot yourself in the foot. All you have to do is change your ```yapActionItems``` method, and forget to change either of these 2 methods. (I'm speaking from experience here :) )

So I think the change I want to make is to scrap these 2 methods, and replace them with a cache (or something like that) (within YapDatabaseActionManager itself).

Long story short:
- I want to improve the external API to make things easier
- But this won't really affect you, because you'll just end up deleting a few methods (less code!)

**Comment by davidchiles** [comment]
@robbiehanson Thanks for the update.

I've started using ActionManager in ChatSecure. I was thinking it would be nice to have a `YapDatabaseReadWriteTransaction` in the `YapActionItemBlock`. 

I may not be using it  exactly how I intended it but in [our case](https://github.com/ChatSecure/ChatSecure-iOS/blob/0fd51f44af28a63997873c00af70fefed3282289/ChatSecure/Classes/Model/Yap%20Storage/YapActions/BuddyActions.swift) once I complete an action I need to remove it from the database so it doesn't fire again. Let me know if I'm making an obvious mistake here.

**Comment by robbiehanson** [comment]
Good question.

So the YapActionItemBlock isn't fired within the context of a transaction. They are all executed on a background queue:
https://github.com/yapstudios/YapDatabase/blob/experimental/action-manager/YapDatabase/Extensions/ActionManager/YapDatabaseActionManager.m#L675

The reason for this is that many actions will have no need for a transaction. For example, an action that needs to refresh data from the server won't need a transaction. At least not until the HTTP response comes back.

> once I complete an action I need to remove it from the database so it doesn't fire again

This is the correct thing to do. The implicit contract is that your YapActionItemBlock will do "something", and then update the database afterwards in such a manner that will either remove the YapActionItem, or update it to fire again at some future date.

(And as mentioned above, the YapActionManager makes no assumptions about how long your ActionItemBlock will take. It might do something asynchronously, and ultimately take several minutes before it completes & updates the database. And this is completely fine. The YapActionManager marks these actionItems as started/executing, and has rules not to fire them again. And since YapActionManager executes these asynchronously on a background queue, the code within an ActionItemBlock won't block the YapActionManager itself.)

**Comment by robbiehanson** [comment]
Update on this issue:

I did a bit of refactoring of YapDatabaseActionManager. As I mentioned I wanted to do, I killed these methods in the YapActionable protocol:
```objective-c
- (BOOL)hasYapActionItems;

- (NSDate *)earliestYapActionItemDate;
```
This simplifies things quite a bit, as you now only have to implement a single method in order to use the ActionManager.

I also refactored the class to be a standard YapDatabaseExtension. (It's now actually a subclass of YapDatabaseView.) So you register it as you would any other extension.

Also I added an internal cache, so this may give a little performance boost as well.

This is now slated for release in the next major version release.

**Comment by chrisballinger** [comment]
Woo hoo! Awesome work man!

**Comment by davidchiles** [comment]
Nice! ActionManager as a standard YapDatabaseExtension is much easier to use. 

I'm wondering how you would approach this problem:

In our case messages, after they're saved, need to go through 2 actions, encrypting and sending. But the sending actions need to happen in order. So that the first message saved is also the first message sent.

Is there a way to enforce an order to the queue to make sure actions happen in a specific order? Or is that not the responsibility of the ActionManger or ActionItem?

**Comment by robbiehanson** [comment]
> is that not the responsibility of the ActionManger

Good question. I don't think it's really the responsibility of the ActionManager, Now, that being said, the ActionManager will invoke the ActionItems in the order in which they're sorted. However, it will invoke them via dispatch_async onto a global concurrent queue. (Meaning there's no guarantee as to which will get CPU time first.) So... it would be relatively easy to tweak the ActionItem class to accept a dispatch_queue_t parameter. And then the ActionManager would use this queue, if given, instead of a global queue.

However, I don't think this is the answer to your problem. (Regardless, we might add the feature anyway because it's not a bad idea.) But hear me out:

What you're probably looking for is a "queue of encrypted messages ready to be sent", that you can enumerate at any time. In other words, if you're disconnected from the network, then encryption may be possible, but not sending. So, if you relied purely on the ActionItemBlock, then you'd ultimately end up writing your own in-memory queue to store the ready-to-send messages when you're offline. And then writing more code to drain this queue when you come back online. See what I'm saying?

(Also, sending may not be enough to remove the item from the queue. Perhaps you require a server ack...)

So, as an alternative, you could create a dedicated YapDatabaseView for this task. If you have multiple accounts, then you could have a group for each account. And the view holds the (sorted) list of messages that are queued for sending. So then you simply have a "SendMessageManager", which listens for YapDatabaseModifiedNotification(s). When it gets a notification it can check to see if that view was modified. And, if so, then it begins the process of draining the queue (sending the messages if account is online).

(And if you support server acks, then the process is similar, but the SendMessageManager also tracks which messages have been sent but un-acked.)

**Comment by Ji11erBug** [comment]
Is there any example how(where) to instantiate and register YapDatabaseActionManager?

**Comment by robbiehanson** [comment]
Generally you would register this extension when you're setting up your database (along with all other extensions that you register at app launch). Here's a simple example:

```objective-c
YapDatabaseActionManager *actionManager = [[YapDatabaseActionManager alloc] init];
	
NSString *const extName = @"ActionManager";
[database asyncRegisterExtension:actionManager withName:extName completionBlock:^(BOOL ready) {
	
	if (!ready) {
		DDLogError(@"Error registering \"%@\" !!!", extName);
	}
}];
```

If you want to get more advanced, there are a few other things you can tweak. One is that it will accept YapDatabaseViewOptions as an init option:

```objective-c
@interface YapDatabaseActionManager : YapDatabaseView

- (instancetype)init;
- (instancetype)initWithOptions:(nullable YapDatabaseViewOptions *)options;
```

This means that you can give it a [YapWhitelistBlacklist](https://github.com/yapstudios/YapDatabase/blob/master/YapDatabase/Utilities/YapWhitelistBlacklist.h) of collections that it should include/ignore.

Also, you can configure it's reachability instance:

```objective-c
/**
 * YapDatabaseActionManager relies on a reachability instance to monitory for internet connectivity.
 * This is to support the YapActionItem.requiresInternet property.
 * 
 * If an instance is not assigned, then one will be automatically created (after registration)
 * via [Reachability reachabilityForInternetConnection].
**/
@property (atomic, strong, readwrite, nullable) Reachability *reachability;
```

So, if you already have a reachability instance that you're using within your app, then you can tell YDBActionManager to use the same instance.

Let me know if you have other questions.

**Comment by Ji11erBug** [comment]
Should we use ActionManger view as persistent or not?

**Comment by robbiehanson** [comment]
Yes, you'll want it to be persistent.

**Comment by robbiehanson** [comment]
The YapDatabaseActionManager has been merged into the v2.9 branch, which is slated for release soon.

**Comment by davidchiles** [comment]
YapDatabaseActionManager is awesome and we're also using it in ChatSecure for some things. But I also made [YapTaskQueue](https://github.com/davidchiles/YapTaskQueue) which works a little different but takes a lot of inspiration from YapDatabaseActionManager. In this case each `queue

**Comment by Ji11erBug** [comment]
I think there should be initializer with versionTag and options for ActionManager.
